### PR TITLE
Refactor 3d calibration

### PR DESCRIPTION
**Pull request by ChingT**
- Remove C++ dependency; Replace with Scipy function for bundle adjustment optimization in 3d calibration (used in `calibrate_3d_binocular`, `calibrate_3d_monocular`, `HMD_Calibration_3D`).
- Move data processing related functions from `calibrate.py`, `finish_calibration.py` and `hmd_calibration.py` to `data_processing.py`.
- Move the calibration related code from `finish_calibration.py` and `hmd_calibration.py` to `calibration_methods.py`.
- Move the actual 3d calibration optimization related functions from `finish_calibration.py` and `hmd_calibration.py` to `calibrate_3d.py`.
- Remove unused functions in `calibrate.py`.
- Rename `calibrate.py` as `calibrate_2d.py`.
- Correct few typos.

Note: assumption for the calibration: 3d eye models are not updated during the calibration time.
TODO: ensure the assumption.

### CODE REVIEW

File: pupil_src/shared_modules/calibration_routines/data_processing.py [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+def filter_confidence(pupil_list, threshold):
+    len_pre_filter = len(pupil_list)
+
+    pupil_list = [p for p in pupil_list if p["confidence"] >= threshold]
+
+    len_post_filter = len(pupil_list)
+    try:
+        dismissed_percentage = 100 * (1.0 - len_post_filter / len_pre_filter)
+    except ZeroDivisionError:
+        pass  # empty pupil_list, is being handled in match_data
```
Comment (pfaion) [inline pupil_src/shared_modules/calibration_routines/data_processing.py]: I feel like this should return an empty list in this case for consistency.

Comment (pfaion) [reply pupil_src/shared_modules/calibration_routines/data_processing.py]: You could even catch this case at the beginning:
```python
def filter_confidence(pupil_list, threshold):
    if not pupil_list:
        return pupil_list
    # ...
```
And get rid of all the try-catch-else stuff.

File: pupil_src/shared_modules/calibration_routines/data_processing.py [hunk 2]
```diff
@@ -0,0 +1,203 @@
...
+    matched_pupil0_data = closest_matches_monocular(ref_list, pupil0)
+    matched_pupil1_data = closest_matches_monocular(ref_list, pupil1)
+
+    if len(matched_pupil0_data) > len(matched_pupil1_data):
+        matched_monocular_data = matched_pupil0_data
+    else:
+        matched_monocular_data = matched_pupil1_data
+
+    logger.info(
+        "Collected {} monocular calibration data.".format(len(matched_monocular_data))
```
Comment (pfaion) [inline pupil_src/shared_modules/calibration_routines/data_processing.py]: f-strings are awesome 😉 
No need to change all `.format()` things now, but keep in mind for the future, I thinks it's more readable like this.

```suggestion
        f"Collected {len(matched_monocular_data)} monocular calibration data."
```

File: pupil_src/shared_modules/calibration_routines/data_processing.py [hunk 3]
```diff
@@ -0,0 +1,203 @@
...
+            closest_p0_idx = _find_nearest_idx(pupil0_ts, r["timestamp"])
+            closest_p0 = pupil0[closest_p0_idx]
+            dispersion = max(closest_p0["timestamp"], r["timestamp"]) - min(
+                closest_p0["timestamp"], r["timestamp"]
+            )
+            if dispersion < max_dispersion:
+                matched.append({"ref": r, "pupil": closest_p0})
+            else:
+                pass
+    return matched
```
Comment (pfaion) [inline pupil_src/shared_modules/calibration_routines/data_processing.py]: The `0` seems to hint at this being eye0-data, but you are calling this with both data from eye0 and eye1, correct?

**Comment by pfaion** [comment]
@ChingT did you test all changed calibrations, also HMD?

**Comment by ChingT** [comment]
@pfaion yes, calibrations in all cases have been tested.

| Binocular | 3d accuracy | 3d precision | 2d accuracy | 2d precision |
| :---: | -------- | -------- |-------- | -------- |
| master | 2.47°<br/> (1856 / 2263) | 0.15°<br/> (**1412** / 2262) | 0.48°<br/> (1856 / 2263) | 0.13°<br/> (1436 / 2262) |
| calibration-3d-refactor | 2.47°<br/> (1856 / 2263) | 0.15°<br/> (**1420** / 2262) | 0.48°<br/> (1856 / 2263) | 0.13°<br/> (1436 / 2262) |

| Monocular | 3d accuracy | 3d precision | 2d accuracy | 2d precision |
| -------- | -------- | -------- |-------- | -------- |
| master | **2.82°**<br/> (796 / 989) | 0.12°<br/> (941 / 988) | 0.24°<br/> (989 / 989) | 0.15°<br/> (954 / 988) |
| calibration-3d-refactor | **2.83°**<br/> (796 / 989) | 0.12°<br/> (941 / 988) | 0.24°<br/> (989 / 989) | 0.15°<br/> (954 / 988) |

- The change of the accuracy/precision introduced by the proposed 3d calibration optimization is negligible.
- The refactor of the code doesn't affect the 2d calibration.

**Comment by papr** [comment]
@pfaion Please review again and merge if you are happy. This is important for the gazer refactor.

**Comment by pfaion** [comment]
@papr I just tested the branch again locally and got a crash first. It worked when I deleted the cpp stuff, shared library and __pychache__ (not sure which of those was responsible).

We need to remember that we do that on the bundle VMs and we should mention that specifically to users building from source!

**Comment by pfaion** [comment]
@papr Should we include all necessary changes to the docs as well before merging? Or do we make another PR for upgrading docs, deployment etc?

**Comment by papr** [comment]
I would suggest a separate pr before merging to master.

File: pupil_src/shared_modules/calibration_routines/optimization_calibration/calibrate_3d.py [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+
+    # Find initial guess for the poses in eye coordinates
+    initial_rotation_matrix, _ = utils.find_rigid_transform(
+        unprojected_ref_points, pupil_normals
+    )
+    hardcoded_translation = (
+        eye0_hardcoded_translation if pupil_id == 0 else eye1_hardcoded_translation
+    )
+    initial_rotation = cv2.Rodrigues(initial_rotation_matrix)[0].ravel()
+    initial_translation = -np.dot(initial_rotation_matrix, hardcoded_translation)
```
Comment (papr) [inline pupil_src/shared_modules/calibration_routines/optimization_calibration/calibrate_3d.py:122]: @ChingT Why is the dot product applied to this initial translation, but not in the binocular case? Has this something to do with the fact that you pass (0, 0, 0) below?

**Review by papr — COMMENTED** [review]
CU https://app.clickup.com/t/pqe6hy

**Pull request merged**